lib/cc/analyzer/engine.rb

```diff
@@ -35,32 +34,30 @@ def run(stdout_io, stderr_io = StringIO.new)
           end
         end
 
-        Thread.new do
-          sleep TIMEOUT
-
-          if engine_running
-            Thread.current.abort_on_exception = true
-            run_command("docker kill #{container_name}")
-
-            Analyzer.statsd.increment("cli.engines.result.error")
-            Analyzer.statsd.increment("cli.engines.result.error.timeout")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error.timeout")
-            raise EngineTimeout, "engine #{name} ran past #{TIMEOUT} seconds and was killed"
+        begin
+          status = nil
+          Timeout::timeout(TIMEOUT) do
+            pid, status = Process.waitpid2(pid)
           end
+        rescue Timeout::Error
+          run_command("docker kill #{container_name} || true")
+          timed_out = true
```

Comment:
All good suggestions. Just waiting for someone (@brynary) to tell us why this is a bad idea first.